src/liballoc/pin.rs

```diff
@@ -0,0 +1,304 @@
+// Copyright 2012-2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Types which pin data to its location in memory
+//!
+//! It is sometimes useful to have objects that are guaranteed to not move,
+//! in the sense that their placement in memory does not change, and can thus be relied upon.
+//!
+//! A prime example of such a scenario would be building self-referencial structs,
+//! since moving an object with pointers to itself will invalidate them,
+//! which could cause undefined behavior.
+//!
+//! In order to prevent objects from moving, they must be *pinned*,
+//! by wrapping the data in special pointer types, such as [`PinMut`] and [`PinBox`].
+//! On top of ensuring the data cannot be taked by value by being pointers,
+//! these types restrict access to the underlying data such that it cannot be moved out of them,
+//! unless the type implements the [`Unpin`] trait,
```

Comment:
Please end the sentence after "out of them" (it's getting pretty long) and start a new one for `Unpin`. `Unpin` deserves more than half a sentence. It should also mention that for `Unpin` types, we have `PinMut<T> = &mut T` and `PinBox<T> = Box<T>`, i.e. the pinned references "degenerate".